### DISCUSSION TITLE
Should I store function references in Zustand?

### DISCUSSION THREAD

**Discussion by ccpu** [pmndrs/zustand#1579] [Q&A] [answered]
According to an [answer on StackOverflow](https://stackoverflow.com/a/49699868/1103757), it is not possible to store function references in Redux, as the store state needs to be serializable.

I am wondering if this is true for the Zustand library as well, or does it not matter if I store function references in the store?

**Comment by dai-shi** [comment, marked as answer]
You can store functions. All those actions in readme are functions.

**Reply by khuongtp** [reply]
Hi @dai-shi , can I ask that if I store a function and then mutate it somewhere, do the consumers re-render? Because it seems that they don't

**Reply by dai-shi** [reply]
It depends on how it's consumed, but I think they should. Can you share some code snippets?

**Reply by khuongtp** [reply]
I'm having a d3 axis (it's a function) that stored like this 
```ts
if (!storeRef.current) {
  storeRef.current = createStore((set) => {
    return {
      xScale1: undefined,
      setChartState: ({ key, value }) => {
        set((state) => {
          return {
            ...state,
            [key]: value,
          };
        });
      },
    };
  });
}
```
Then, somewhere I call setNewScale like this
```ts
const xScale = useChartStore((state) => {
  return state.xScale1;
});
const setNewScale = () => {
  const newXScale = (xScale as ScaleBand<string>).range([
    0,
    width - marginLeft,
  ]);
  setChartState({
    key: "xScale1",
    value: newXScale,
  });
};
```
Here's the hook
```ts
export const useChartStore = ((selector) => {
  const store = useContext(ChartContext);
  if (!store) {
    throw new Error("Missing StoreProvider");
  }
  return useStore(store, selector);
}) as UseBoundStore<
  StoreApi<ChartState & { setChartState: SetState<ChartState> }>
>;
```
The consumers just don't re-render

**Reply by dai-shi** [reply]
I thought you wanted to mutate the function itself, but it seems like your use case is a normal one.

Probably it's because of stale closures. How about function update:
```js
if (!storeRef.current) {
  storeRef.current = createStore((set) => {
    return {
      xScale1: undefined,
      setChartState: ({ key, update }) => {
        set((state) => {
          return {
            ...state,
            [key]: update(state[key]),
          };
        });
      },
    };
  });
}
```

**Reply by khuongtp** [reply]
Turns out d3 axis has a `copy` method and it works for my case 😃 , thank you for your quick replies tho!
